documentation/memory/step3/using_gc_traces.md

```diff
@@ -20,4 +20,80 @@ v8.setFlagsFromString('--trace_gc');
 setTimeout(() => { v8.setFlagsFromString('--notrace_gc'); }, 60e3);
 ```
 
-## Examining a trace
\ No newline at end of file
+## Examining a trace
+
+Obtained traces of garbage collection looks like the following lines.
+
+```
+[19278:0x5408db0]  44 ms: Scavenge 2.3 (3.0) -> 1.9 (4.0) MB, 1.2 / 0.0 ms  (average mu = 1.000, current mu = 1.000) allocation failure
+
+[23521:0x10268b000]  120 ms: Mark-sweep 100.7 (122.7) -> 100.6 (122.7) MB, 0.15 / 0.0 ms  (average mu = 0.132, current mu = 0.137) deserialize GC in old space requested
+```
+
+This is how to interpret the trace data:
+
+<table>
+  <tr>
+    <th>Token value</th>
+    <th>Interpretation</th>
+  </tr>
+  <tr>
+    <td>23521</td>
+    <td>PID of the running process</td>
+  </tr>
+  <tr>
+    <td>0x10268db0</td>
+    <td>Isolate (JS heap instance)</td>
+  </tr>
+  <tr>
+    <td>120</td>
+    <td>Time since the process start in ms</td>
+  </tr>
+  <tr>
+    <td>Mark-sweep</td>
+    <td>Type / Phase of GC</td>
+  </tr>
+  <tr>
+    <td>100.7</td>
+    <td>Heap used before GC in MB</td>
+  </tr>
+  <tr>
+    <td>122.7</td>
+    <td>Total heap before GC in MB</td>
+  </tr>
+  <tr>
+    <td>100.6</td>
+    <td>Heap used after GC in MB</td>
+  </tr>
+  <tr>
+    <td>122.7</td>
+    <td>Total heap after GC in MB</td>
+  </tr>
+  <tr>
+    <td>0.15 / 0.0 </td>
+    <td>Time spent in GC in ms</td>
+  </tr>
+  <tr>
+    <td>(average mu = 0.132, current mu = 0.137) deserialize GC in old space requested</td>
+    <td>Reason for GC</td>
+  </tr>
+</table>
+
+## Examples of diagnosing memory issues with trace option:
+
+A. How to get context of bad allocations using --trace-gc
+  1. Suppose we observe that the old space is ocntinously increasing.
+  2. But due to heavy gc, the heap roof is not hit, but the process is slow.
+  3. Review the trace data and figure out how much is the total heap before and after the gc.
+  4. Reduce `--max-old-space-size` such that the total heap is closer to the limit.
+  5. Allow the program to run, hit the out of memory.
+  6. The produced log gives shows the failing context.
+
+B. How to assert whether too many gc are happening or too many gc is causing an overhead
+  1. Review the trace data, specifically around time between consecutive gcs.
+  2. Review the trace data, specifically around time spent in gc.
+  3. If the time between two gc is less than the time spent in gc, the application is sseverely starving.
+  4. If the time between two gc and the time spent in gc are very high, probably the application can use a smaller heap
+  5. If the time between two gc is much greater than the time spent in gc, application is relatively healthy
+  6. While the actual numbers for these metrics change from workload to workload, a reasonable gap between gcs is 20 minutes, and a reasonable gc time is < 100 ms.
```

Comment:
Not sure if we should mention concrete times here as it heavily depends on the application and the workload.
In special the 20minues seems to be quite long to me.

Reply:
Removed the last step.

Reply:
I had the same thought as well.  20 minutes seems to long. Do we have something to back up suggestions for particular numbers?